.github/workflows/docker-build.yml

```diff
@@ -0,0 +1,26 @@
+name: Docker image build
+
+on: [push, release]
+
+jobs:
+  docker-build-push:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
```

Comment:
I don't quite understand what this does. [The documentation][1] says it “will create and boot a builder”. What does that mean? The whole pipeline is running on a GitHub-supplied builder. Why do we need to boot another builder? Who will provide it?

I found [a Docker blog post][2] that seems to imply that the whole point of this Buildx thing is to cache image layers so that we don't need to download them on each run. But this only works if used in conjunction with actions/cache. However, we are already using actions/cache for caching ccache data (see the existing workflow). Could this lead to a conflict in cache management?

It may be worth noting that the GitHub runners come with [docker and podman pre-installed][3].

[1]: https://github.com/marketplace/actions/docker-setup-buildx
[2]: https://www.docker.com/blog/docker-github-actions/
[3]: https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md#tools

Reply:
As I understand, docker buildx is a plugin to docker. It is necessary for multi-arch builds, else this is not necessary. Booting builder means probably means starting a container. I am not sure how buildx works fully well, I can see that it uses a docker image `moby/buildkit` containing BuildKit, essentially running the build itself in a temporary container. This also slows down the build a bit since the built image has to be transferred to the 'outside' over docker network.

This too can be removed if multi-arch build is not a requirement.

>However, we are already using actions/cache for caching ccache data (see the existing workflow). Could this lead to a conflict in cache management?

I think using cache requires restoration of cache from the `actions/cache@v3`. Which is not being done in the docker build. I think that cache can be used here by running the same action. But this requires a further step in Dockerfile (copying of cache files from environment to root of the build), which is not compatible with building outside GitHub workflows. This might need extra logic (like looking for environment variables present in GitHub environment) or using a different Dockerfile altogether.

Also, layer caching is also not useful here for the build step as it isn't intelligent as ccache. Layer caching mechanism uses hash of input files and the step command for each step. If any changes are made, then the whole layer cache is invalidated. So the only useful layers would be until the setup and the build step can never be cached.

This is the reason to copy only `install-dependencies.sh` in the beginning, or all the layers would be rebuilt.